src/rules/selector-max-depth/index.js

```diff
@@ -0,0 +1,78 @@
+import resolvedNestedSelector from "postcss-resolve-nested-selector"
+
+import {
+  isStandardRule,
+  isStandardSelector,
+  report,
+  ruleMessages,
+  validateOptions,
+} from "../../utils"
+
+export const ruleName = "selector-max-depth"
+
+export const messages = ruleMessages(ruleName, {
+  expected: (selector, depth) => `Expected "${selector}" to have a depth equal to or less than ${depth}`,
+})
+
+export default function (max) {
+  return (root, result) => {
+    const innerSelectorReg = /(?::not\((.*?)\))/g
+    const validOptions = validateOptions(result, ruleName, {
+      actual: max,
+      possible: [function (max) {
+        return typeof max === "number" && max > 0
+      }],
+    })
+    if (!validOptions) { return }
+
+    // Compares depth of a selector with the maximum
+    function checkDepth(selector, rule) {
```

Comment:
I was thinking of that, but there are several arguments against that:

1. `postcss-selector-parser` uses regexes too;
2. The docs and the API differ from the version used by stylelint (e.g, `walkTags` in [the docs](https://github.com/postcss/postcss-selector-parser/blob/master/API.md), but `eachTag` in the actual code of v1.3.3)
3. It's slower:
    ```
    npm run benchmark-rule selector-max-depth 3
    
    Warnings: 279
    Mean: 152.12407722222224 ms
    Deviation: 27.9745142792245 ms
    ```
    
    Maybe it was me who used it all wrong (though the api is quite simple - just traversing stuff).